# TASK 1



In [23]:
#importing the required libraries
import pandas as pd
import numpy as np

In [24]:
#mounting Google Drive in order to access the datasets for the 2 tasks
from google.colab import drive
drive .mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [25]:
#loading the dataset 'hotel.csv' stored in 'task' folder of Google Drive into 'data'
data=pd.read_csv("/content/drive/My Drive/task/hotel.csv") 

In [26]:
#displaying the entire dataset (this dataset is a dummy dataset prepared by me with 3 columns having dates)
data

S. No First Name Last Name  ...                 Address     Arrival   Departure
0       1     Ashish    Tiwari  ...     Varanasi, UP, India  02-02-2020   5-02-2020
1       2      Zadir     Ahmed  ...     Jorhat, Assam India      4/2/20  11-02-2020
2       3   Priyanka   Parasar  ...  Guwahati, Assam, India  11-02-2020     15/2/20
3       4     Robert     Smith  ...   Greenwich, London, UK  12-02-2020  13-02-2020
4       5       Lana     White  ...         Ontario, Canada  19-02-2020  23-02-2020
5       6       John  Williams  ...     Fremont, California  20-02-2020    22-02-20
6       7        Ram    Mishra  ...           Shimla, India   21/2/2020  27-02-2020
7       8       Rana     Singh  ...        Ajmer, Rajasthan  21-02-2020  23-02-2020
8       9    Kuldeep     Kumar  ...  Indore, Madhya Pradesh  23-02-2020  01-03-2020
9      10       Jane  Atkinson  ...    Melbourne, Australia  03-03-2020     10/3/20
10     11   Jennifer   Perkins  ...              Lima, Peru  16-03-2020   17/3/2020
11     12        Sam    Rogers  ...     La Plata, Argentina   16/3/2020  20-03-2020
12     13      Ayush     Bhatt  ...      Ahmedabad, Gujarat  28-03-2020  03-04-2020
13     14      Pooja    Sharma  ...        Dibrugarh, Assam      5/4/20  13-04-2020
14     15     Souvik    Baruah  ...        New Delhi, India  14-04-2020  16-04-2020
15     16      Rohit     Bajaj  ...          Bandra, Mumbai  20-04-2020  22-04-2020
16     17     Shreya    Pandey  ...         Andheri, Mumbai  29-04-2020    6/5/2020
17     18        Rex      Ross  ...  Pretoria, South Africa  06-05-2020    9/5/2020
18     19      Payal     Dubey  ...           Tezpur, Assam  19-05-2020  20-05-2020
19     20    Farhina     Ahmed  ...         Sivsagar, Assam  01/06/2020  03-06-2020
20     21     Aritra   Kaushik  ...          Garmur, Jorhat  18-05-2020  23-06-2020
21     22     Pijush    Bhuyan  ...      Bikaner, Rajasthan  19-05-2020  21-05-2020
22     23    Charles    Wilson  ...             Queens, NYC  27-05-2020  28-05-2020
23     24      James     Moore  ...           Brooklyn, NYC  08-06-2020  11-06-2020
24     25      Brent    Rivera  ...              Dubai, UAE  10-06-2020  12-06-2020

[25 rows x 7 columns]

From the above output we can clearly see that the columns 'DOB', 'Arrival' and 'Departure' have dates stored in them. In order to verify that the values stored by these columns are dates we have to check whether the datatype of these three columns is 'object'. The reason for this is explained below after checking the datatypes of the three columns.

In [27]:
# printing the first few rows of the columns 'DOB', 'Arrival' and 'Departure'
print(data['DOB'].head())
print(data['Arrival'].head())
print(data['Departure'].head())

0    02/11/1999
1      10-03-92
2    03-13-1986
3       12/2/99
4    23-01-2000
Name: DOB, dtype: object
0    02-02-2020
1        4/2/20
2    11-02-2020
3    12-02-2020
4    19-02-2020
Name: Arrival, dtype: object
0     5-02-2020
1    11-02-2020
2       15/2/20
3    13-02-2020
4    23-02-2020
Name: Departure, dtype: object


We can see that the datatypes of the columns 'DOB', 'Arrival' and 'Departure' have been shown as 'object'. Pandas uses this datatype to refer mainly to strings. As dates are nothing but strings, the output 'dtype: object' verifies that these columns have dates stored in them.

In [28]:
# checking the data type of 'DOB' column without printing the first few rows
data['DOB'].dtype

dtype('O')

In [29]:
# checking the data type of 'Arrival' column without printing the first few rows
data['Arrival'].dtype

dtype('O')

In [30]:
# checking the data type of 'Departure' column without printing the first few rows
data['Departure'].dtype

dtype('O')

We can see that the datatypes of the columns 'DOB', 'Arrival' and 'Departure' are shown as 'O' where 'O' is nothing but the code for 'object'.

In [31]:
#Since, Pandas doesn't automatically recognise the date values as 'date', therefore, we have to convert them so that they are recognized as dates.
# Creating a new column, DOB_1, with the parsed dates 
# Handling different date formats present in the dataset and converting them into the same format 'year-month-day'
data['DOB_1'] = pd.to_datetime(data['DOB'], infer_datetime_format=True)
data['DOB_1'].head()

0   1999-02-11
1   1992-10-03
2   1986-03-13
3   1999-12-02
4   2000-01-23
Name: DOB_1, dtype: datetime64[ns]

In [32]:
# creating a new column, Arrival_1, with the parsed dates
# Handling different date formats present in the dataset and converting them into the same format 'year-month-day'
data['Arrival_1'] = pd.to_datetime(data['Arrival'], infer_datetime_format=True)
data['Arrival_1'].head()

0   2020-02-02
1   2020-04-02
2   2020-11-02
3   2020-12-02
4   2020-02-19
Name: Arrival_1, dtype: datetime64[ns]

In [33]:
# creating a new column, Departure_1, with the parsed dates
# Handling different date formats present in the dataset and converting them into the same format 'year-month-day'
data['Departure_1'] = pd.to_datetime(data['Departure'], infer_datetime_format=True)
data['Departure_1'].head()

0   2020-05-02
1   2020-11-02
2   2020-02-15
3   2020-02-13
4   2020-02-23
Name: Departure_1, dtype: datetime64[ns]

We can clearly see that now when we check the datatypes of our new columns 'DOB_1', 'Arrival_1' and 'Departure_1' we get the result as 'datetime64'. They are now specifically recognized as dates.

In [34]:
# printing first few rows of our dataset after parsing the dates
data.head()

S. No First Name Last Name  ...      DOB_1  Arrival_1 Departure_1
0      1     Ashish    Tiwari  ... 1999-02-11 2020-02-02  2020-05-02
1      2      Zadir     Ahmed  ... 1992-10-03 2020-04-02  2020-11-02
2      3   Priyanka   Parasar  ... 1986-03-13 2020-11-02  2020-02-15
3      4     Robert     Smith  ... 1999-12-02 2020-12-02  2020-02-13
4      5       Lana     White  ... 2000-01-23 2020-02-19  2020-02-23

[5 rows x 10 columns]

After this, we print the difference of our newly parsed date columns taking two at a time.

In [35]:
# storing the difference between 'DOB_1' and 'Arrival_1' in a new column name 'DOB_1-Arrival_1'
data['DOB_1-Arrival_1'] = (data['DOB_1'].sub(data['Arrival_1'], axis=0)).abs()
data.head()

S. No First Name Last Name  ...  Arrival_1 Departure_1 DOB_1-Arrival_1
0      1     Ashish    Tiwari  ... 2020-02-02  2020-05-02       7661 days
1      2      Zadir     Ahmed  ... 2020-04-02  2020-11-02      10043 days
2      3   Priyanka   Parasar  ... 2020-11-02  2020-02-15      12653 days
3      4     Robert     Smith  ... 2020-12-02  2020-02-13       7671 days
4      5       Lana     White  ... 2020-02-19  2020-02-23       7332 days

[5 rows x 11 columns]

In [36]:
# storing the difference between 'Arrival_1' and 'Departure_1' in a new column name 'Arrival_1-Departure_1'
data['Arrival_1-Departure_1'] = (data['Arrival_1'].sub(data['Departure_1'], axis=0)).abs()
data.head()

S. No First Name Last Name  ... Departure_1 DOB_1-Arrival_1 Arrival_1-Departure_1
0      1     Ashish    Tiwari  ...  2020-05-02       7661 days               90 days
1      2      Zadir     Ahmed  ...  2020-11-02      10043 days              214 days
2      3   Priyanka   Parasar  ...  2020-02-15      12653 days              261 days
3      4     Robert     Smith  ...  2020-02-13       7671 days              293 days
4      5       Lana     White  ...  2020-02-23       7332 days                4 days

[5 rows x 12 columns]

In [37]:
# storing the difference between 'Departure_1' and 'DOB_1' in a new column name 'Departure_1-DOB_1'
data['Departure_1-DOB_1'] = (data['Departure_1'].sub(data['DOB_1'], axis=0)).abs()
data.head()

S. No First Name  ... Arrival_1-Departure_1 Departure_1-DOB_1
0      1     Ashish  ...               90 days         7751 days
1      2      Zadir  ...              214 days        10257 days
2      3   Priyanka  ...              261 days        12392 days
3      4     Robert  ...              293 days         7378 days
4      5       Lana  ...                4 days         7336 days

[5 rows x 13 columns]

In [38]:
# printing the entire dataset after all the above modifications
data

S. No First Name  ... Arrival_1-Departure_1 Departure_1-DOB_1
0       1     Ashish  ...               90 days         7751 days
1       2      Zadir  ...              214 days        10257 days
2       3   Priyanka  ...              261 days        12392 days
3       4     Robert  ...              293 days         7378 days
4       5       Lana  ...                4 days         7336 days
5       6       John  ...                2 days         7879 days
6       7        Ram  ...                6 days        12110 days
7       8       Rana  ...                2 days         6656 days
8       9    Kuldeep  ...               51 days               NaT
9      10       Jane  ...              214 days        18737 days
10     11   Jennifer  ...                1 days        10904 days
11     12        Sam  ...                4 days        13563 days
12     13      Ayush  ...               24 days         8584 days
13     14      Pooja  ...               21 days        13806 days
14     15     Souvik  ...                2 days         7715 days
15     16      Rohit  ...                2 days               NaT
16     17     Shreya  ...               37 days         7917 days
17     18        Rex  ...               92 days        10085 days
18     19      Payal  ...                1 days         8063 days
19     20    Farhina  ...               60 days         7592 days
20     21     Aritra  ...               36 days         7113 days
21     22     Pijush  ...                2 days        19970 days
22     23    Charles  ...                1 days               NaT
23     24      James  ...               92 days        22274 days
24     25      Brent  ...               61 days        18899 days

[25 rows x 13 columns]

# TASK 2


In [39]:
# loading the dataset 'pokemon.csv' (publicly available dataset collected from Kaggle) stored in the 'task' folder of Google Drive into 'data'
data=pd.read_csv("/content/drive/My Drive/task/pokemon.csv")
#displaying first few rows of our dataset 
data.head()

Number        Name Type_1  ... Weight_kg  Catch_Rate      Body_Style
0       1   Bulbasaur  Grass  ...       6.9          45       quadruped
1       2     Ivysaur  Grass  ...      13.0          45       quadruped
2       3    Venusaur  Grass  ...     100.0          45       quadruped
3       4  Charmander   Fire  ...       8.5          45  bipedal_tailed
4       5  Charmeleon   Fire  ...      19.0          45  bipedal_tailed

[5 rows x 23 columns]

In [40]:
#checking the number of rows and columns present in our dataset
data.shape

(721, 23)

We see that our dataset contains a total of 721 rows and 23 columns. Now, the task is to drop all the columns having Pearson correlation more than 0.85.

In [41]:
# Creating the correlation matrix
corr_matrix = data.corr().abs()

# Selecting the upper triangle of the correlation matrix
upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(np.bool))

# Finding the index of the feature columns with Pearson correlation greater than 0.85
to_drop = [column for column in upper.columns if any(upper[column] > 0.85)]

In [42]:
# Droping the columns having Pearson correlation more than 0.85
# after dropping, we print the dataset again to verify that some of the columns are dropped
data.drop(data[to_drop], axis=1)

Number        Name   Type_1  ... Weight_kg  Catch_Rate       Body_Style
0         1   Bulbasaur    Grass  ...       6.9          45        quadruped
1         2     Ivysaur    Grass  ...      13.0          45        quadruped
2         3    Venusaur    Grass  ...     100.0          45        quadruped
3         4  Charmander     Fire  ...       8.5          45   bipedal_tailed
4         5  Charmeleon     Fire  ...      19.0          45   bipedal_tailed
..      ...         ...      ...  ...       ...         ...              ...
716     717     Yveltal     Dark  ...     203.0          45        two_wings
717     718     Zygarde   Dragon  ...     305.0           3  serpentine_body
718     719     Diancie     Rock  ...       8.8           3        head_arms
719     720       Hoopa  Psychic  ...       9.0           3        head_only
720     721   Volcanion     Fire  ...     195.0           3        quadruped

[721 rows x 22 columns]

So, we can see that the 'Generation' column has been dropped and the total number of columns got reduced to 22 !